File: microservice-architecture/examples/microservice-basics/python/test/test_http.py
import os
import sys
import threading
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from urllib.parse import urlparse, parse_qs

sys.path.append(os.getcwd())

from src.http_inventory_client import HttpInventoryClient
from src.order_service import OrderService


class InventoryHandler(BaseHTTPRequestHandler):
    stock = 2

    def do_GET(self):
        parsed = urlparse(self.path)
        if parsed.path != '/reserve':
            self.send_response(404)
            self.end_headers()
            return

        qs = parse_qs(parsed.query)
        sku = qs.get('sku', [''])[0]
        quantity = int(qs.get('quantity', ['0'])[0])

        if sku == 'SKU-BOOK' and quantity > 0 and InventoryHandler.stock >= quantity:
            InventoryHandler.stock -= quantity
            self.send_response(200)
            self.end_headers()
            self.wfile.write(b'OK')
            return

        self.send_response(409)
        self.end_headers()
        self.wfile.write(b'NO_STOCK')

    def log_message(self, fmt, *args):
        return


def assert_equal(expected, actual, message):
    if expected != actual:
        raise RuntimeError(f"{message} expected={expected} actual={actual}")


def main():
    server = ThreadingHTTPServer(('127.0.0.1', 0), InventoryHandler)
    thread = threading.Thread(target=server.serve_forever, daemon=True)
    thread.start()

    base_url = f"http://127.0.0.1:{server.server_port}"

    try:
        service = OrderService(HttpInventoryClient(base_url))

        success = service.create_order('ORD-2001', 'SKU-BOOK', 1)
        assert_equal('CREATED', success.status, 'http status should be CREATED')

        failed = service.create_order('ORD-2002', 'SKU-BOOK', 2)
        assert_equal('REJECTED', failed.status, 'http status should be REJECTED')

        print('microservice-basics(python/http) tests passed')
    finally:
        server.shutdown()
        server.server_close()


if __name__ == '__main__':
    main()

File: microservice-architecture/microservice-basics/java/test/TestHttp.java
package test;

import com.sun.net.httpserver.HttpServer;
import src.HttpInventoryClient;
import src.Order;
import src.OrderService;

import java.io.IOException;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.net.URI;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;

public class TestHttp {

    private static void assertEquals(String expected, String actual, String message) {
        if (!expected.equals(actual)) {
            throw new RuntimeException(message + " expected=" + expected + " actual=" + actual);
        }
    }

    private static Map<String, String> parseQuery(String query) {
        Map<String, String> params = new HashMap<>();
        if (query == null || query.isBlank()) {
            return params;
        }
        String[] pairs = query.split("&");
        for (String pair : pairs) {
            String[] kv = pair.split("=", 2);
            if (kv.length == 2) {
                params.put(kv[0], kv[1]);
            }
        }
        return params;
    }

    public static void main(String[] args) throws IOException {
        HttpServer server = HttpServer.create(new InetSocketAddress(0), 0);
        int[] bookStock = new int[]{2};

        server.createContext("/reserve", exchange -> {
            URI uri = exchange.getRequestURI();
            Map<String, String> query = parseQuery(uri.getRawQuery());
            String sku = query.getOrDefault("sku", "");
            int qty = Integer.parseInt(query.getOrDefault("quantity", "0"));

            int code;
            String body;
            if ("SKU-BOOK".equals(sku) && qty > 0 && bookStock[0] >= qty) {
                bookStock[0] -= qty;
                code = 200;
                body = "OK";
            } else {
                code = 409;
                body = "NO_STOCK";
            }

            exchange.sendResponseHeaders(code, body.getBytes(StandardCharsets.UTF_8).length);
            try (OutputStream os = exchange.getResponseBody()) {
                os.write(body.getBytes(StandardCharsets.UTF_8));
            }
        });

        server.start();
        String baseUrl = "http://127.0.0.1:" + server.getAddress().getPort();

        try {
            OrderService orderService = new OrderService(new HttpInventoryClient(baseUrl));

            Order success = orderService.createOrder("ORD-2001", "SKU-BOOK", 1);
            assertEquals("CREATED", success.getStatus(), "http status should be CREATED");

            Order failed = orderService.createOrder("ORD-2002", "SKU-BOOK", 2);
            assertEquals("REJECTED", failed.getStatus(), "http status should be REJECTED");

            System.out.println("microservice-basics(java/http) tests passed");
        } finally {
            server.stop(0);
        }
    }
}
